strawberry/fastapi/router.py

```diff
@@ -0,0 +1,196 @@
+                operations = json.loads(multipart_data.get("operations", {}))
+                files_map = json.loads(multipart_data.get("map", {}))
+                data = replace_placeholders_with_files(
+                    operations, files_map, multipart_data
+                )
+            else:
+                return PlainTextResponse(
+                    "Unsupported Media Type",
+                    status_code=status.HTTP_415_UNSUPPORTED_MEDIA_TYPE,
+                )
+
+            try:
+                request_data = parse_request_data(data)
+            except MissingQueryError:
+                return PlainTextResponse(
+                    "No GraphQL query found in the request",
+                    status_code=status.HTTP_400_BAD_REQUEST,
+                )
+
+            result = await self.execute(
+                request_data.query,
+                variables=request_data.variables,
+                context=context,
+                operation_name=request_data.operation_name,
+                root_value=root_value,
+            )
+
+            response_data = await self.process_result(request, result)
+
+            return JSONResponse(response_data, status_code=status.HTTP_200_OK)
+
+        @self.websocket("")
+        async def websocket_endpoint(
+            websocket: WebSocket,
+            context=Depends(context_getter),
+            root_value=Depends(get_root_value),
+        ):
+            async def _get_context():
+                return context
+
+            async def _get_root_value():
+                return root_value
+
+            preferred_protocol = self.pick_preferred_protocol(websocket)
+            if preferred_protocol == GRAPHQL_TRANSPORT_WS_PROTOCOL:
+                await self.graphql_transport_ws_handler_class(
+                    schema=self.schema,
+                    debug=self.debug,
+                    connection_init_wait_timeout=self.connection_init_wait_timeout,
+                    get_context=_get_context,
+                    get_root_value=_get_root_value,
+                    ws=websocket,
+                ).handle()
+            elif preferred_protocol == GRAPHQL_WS_PROTOCOL:
+                await self.graphql_ws_handler_class(
+                    schema=self.schema,
+                    debug=self.debug,
+                    keep_alive=self.keep_alive,
+                    keep_alive_interval=self.keep_alive_interval,
+                    get_context=_get_context,
+                    get_root_value=_get_root_value,
+                    ws=websocket,
+                ).handle()
+            else:
+                await websocket.close(code=4406)
+
+    def pick_preferred_protocol(self, ws: WebSocket) -> Optional[str]:
+        protocols = ws["subprotocols"]
+        intersection = set(protocols) & set(self.protocols)
+        sorted_intersection = sorted(intersection, key=protocols.index)
+        return next(iter(sorted_intersection), None)
```

Comment:
Could the last two lines not be replaced with a check for the empty set and the `min` function?
I mean, there is no reason to sort the entire set (`O(n*log(n))`) if we just want to find the smallest element (`O(n)`)?